### ISSUE TITLE
Populate Google Tag data for existing site using GA4

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#6082]
Based on #6077 and #6080, for the GA4 snippet, a Google Tag ID should be used going forward. While this _usually_ is the same as the GA4 measurement ID, we will need to migrate all sites to have this tag ID determined and stored. While #6081 implements this for the "new user" / "updating configuration" scenarios, we also need to have logic that populates the data for existing sites where GA4 is already active and fully configured.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation Brief

* 

### Test Coverage

* 

## QA Brief

* 

## Changelog entry

*

**Comment by techanvil** [comment]
Hi @nfmohit, this IB is looking good. I would suggest one small change. In the **Test Coverage** section, you have specified the following point:

> - Add test case for the new `sync_google_tag_data()` method.

However, `sync_google_tag_data()` will be a private method, so it's not directly testable. To avoid confusion so the implementer doesn't attempt to directly test a private method (we do this in a handful of places, but it's not recommended), I'd suggest changing to something like "add PHPUnit test coverage for the new changes" or such like.

**Comment by nfmohit** [comment]
Good catch. I've updated the IB. Thank you, @techanvil !

**Comment by techanvil** [comment]
Thanks @nfmohit!

IB :white_check_mark:

**Comment by bethanylang** [comment]
**NOTE**: This issue is blocked by #6081 in Sprint 92.

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, great suggestion regarding the centralized timestamp, I've updated the IB to make use of a `googleTagLastSyncedAt` setting as suggested. I've also created issues #6472, #6473 and #6474 to capture the ideas about the cache API.

The only thing I'm not so sure about is the idea of creating a custom `useSyncGoogleTagSettingsEffect` hook. The `DashboardEntryPoint` component is so small, it seems a bit premature to extract a hook which itself is going to only contain a single action dispatch. I don't think we need it for code reuse and without containing any logic it would just add a layer of indirection to understanding what action is being dispatched. If `DashboardEntryPoint` was one of our overly large components then I'd be more in favour of it, but I don't think it's really worthwhile at this point. I don't think it's really necessary for the sake of moving conditional logic either, as without a clear case for code reuse, we'd just have to refer to two places to understand the flow rather than one, and it seems better to keep the logic within `*syncGoogleTagSettings`, IMHO.

Also, it's worth bearing in mind that the AC does mention that "Eventually, this logic could be removed"...

**Comment by aaemnnosttv** [comment]
Thanks @techanvil ! IB overall reads great to me, just one last question.

One thing I'm thinking about is the format of the `googleTagLastSyncedAt` value in case we ever wanted to use this in PHP too, since PHP references the epoch in seconds where as `Date.now` is in milliseconds. We could potentially convert the integer into seconds when saving, and back to millis when passing to the client, or we could use a datetime string which might be less vulnerable to mistakes in translation or errors in interpretation. What do you think?

**Comment by techanvil** [comment]
Interesting point to consider, @aaemnnosttv.

My thoughts on the seconds/milliseconds aspect are, given the client is currently the only known user of this timestamp and is responsible for defining and using it, I think it would be more appropriate to keep it in milliseconds, it doesn't seem worthwhile to add a conversion step for the client for the sake of not converting it on the server, where it may never be used.

Similarly I don't really think it's worthwhile storing as a datetime string considering the only anticipated usage is to do a numeric comparison on it, and it just adds unnecessary complexity. Although, it would be "fairer" than having to choose seconds/milliseconds as both client/server would then need to parse it :smile: 

What I'd suggest under the circumstances is to simply rename the setting to make it clear the value is stored in milliseconds, e.g. `googleTagLastSyncedAtMilliseconds`. What do you think?

**Comment by aaemnnosttv** [comment]
Thanks @techanvil, that sounds reasonable to me it just reads a bit verbose. How about `googleTagLastSyncedAtMs`? 😄 

I'll make this change to avoid another iteration here as the rest of it LGTM 👍

**Comment by aaemnnosttv** [comment]
IB ✅

**Comment by techanvil** [comment]
> Thanks @techanvil, that sounds reasonable to me it just reads a bit verbose. How about `googleTagLastSyncedAtMs`? smile
> 
> I'll make this change to avoid another iteration here as the rest of it LGTM +1

Thanks @aaemnnosttv, SGTM!

**Comment by mohitwp** [comment]
### QA Update ⚠️ 

@nfmohit 

- Tested on main.
- Followed the QAB and verified the response for` googlesitekit.data.select( 'modules/analytics-4' ).getSettings()` .

**Q-1)** In response Google Tag setting properties (e.g. googleTagID) are showing without any value, is this expected ?

![image](https://user-images.githubusercontent.com/94359491/220569886-31157c9f-fb5a-41f1-b348-8570c97f8f3e.png)

**Q-2)** On enabling 'gteSupport' feature flag Additional permission required Pop up appears. Is this expected ?

![image](https://user-images.githubusercontent.com/94359491/220569362-62353377-41bf-4af4-a3ca-b9b0fceee869.png)

cc @techanvil 

<details>

https://user-images.githubusercontent.com/94359491/220569591-7417aa39-7d7b-483f-8a8f-931a98dca820.mp4

</details>

**Comment by techanvil** [comment]
Thanks for raising this @mohitwp. I've investigated, and it's evident that following the QAB as detailed does not result in the GTE settings being immediately updated. Arguably though, this is following the spec to the letter, but this leads to a question for @felixarntz about whether we should change things to make it behave more intuitively.

Following the QAB, setting up Site Kit with Analytics but without the `gteSupport` feature flag enabled, and then enabling the `gteSupport` flag on the Dashboard, the `*syncGoogleTagSettings()` action is triggered. This results in a request for the GTE settings in order to sync them, however the request returns a `missing_required_scopes` error as the user has not yet granted the `https://www.googleapis.com/auth/tagmanager.readonly` scope.

Due to the way `*syncGoogleTagSettings()` handles this error (i.e. to ignore it, which is in keeping with the spec, in theory), this results in the `googleTagLastSyncedAtMs` timestamp being set to the current time, while the "Additional Permissions Required" modal is displayed. When the user proceeds with and returns from the OAuth flow, `*syncGoogleTagSettings()` is invoked again but bails out as the timestamp has just been set, so as far as the action is concerned, it's just made an attempt to sync settings and should not try again for another hour.

As mentioned this is in keeping with the spec but it raises the question, should we in fact make an exception for the `missing_required_scopes` and _not_ update the timestamp in this case, to ensure the sync runs again fully when the user returns from OAuth, so the GTE settings are updated at this point? Otherwise, despite requesting the scope from the user, no action will be taken until an hour has gone by and they hit the Dashboard again.

It seems to be we probably should do this, but we should also consider that if the user does not grant the scope, maybe is unable to for some reason, we could end up in a situation where the sync attempt & scope request is made on every Dashboard load, which is something we're trying to avoid as per the spec.

@felixarntz, please can you advise on how you want to proceed with this one?

**Comment by felixarntz** [comment]
@techanvil Great catch! Thinking about this further, we may be best off not attempting any of the syncing logic if the user hasn't granted the scope yet.

A few considerations here:
* The scope will only be missing for existing plugin users that previously were using Analytics but not Tag Manager. New users will be prompted to grant both scopes right away, so this is only a problem for existing users (which of course are a lot, so it is definitely important to improve this).
* Since the new Tag Manager scope will not be an Analytics base scope, users will already see a banner notification immediately after activating the feature flag (see #6421), so they should (hopefully) grant the scope very quickly.
* I think we should check before even attempting the sync whether the current user has already granted the scope. If not, we already know ahead of attempting it that it will fail, so in that case we should not attempt it and thus also not set the timestamp that is associated with it.

This also avoids the somewhat ugly experience of being on the dashboard and suddenly seeing this popup without even having invoked any action. Does that sound reasonable and would work well with our implementation?

**Comment by techanvil** [comment]
> This also avoids the somewhat ugly experience of being on the dashboard and suddenly seeing this popup without even having invoked any action. Does that sound reasonable and would work well with our implementation?

Thanks @felixarntz - sounds like a good solution. This was a quick fix, I've created a followup PR and tweaked the QAB accordingly, cc @mohitwp.

**Comment by mohitwp** [comment]
### QA Update ✅  


- Tested on main.
- Followed the QAB and now Google Tag setting properties (e.g. googleTagID) are populated after enabling gteSupport Feature flag.
- The Dashboard now displays a notification to redo the setup for Tag Manager.
- **Note**—I'm not sure from where we are fetching `googleTagAccountID` and `googleTagContainerID`. I verified tag container ID and account ID, but that is different. Confirmed with @nfmohit that it is fetching expected response, might be I'm not looking at right place. 

 cc @techanvil @aaemnnosttv 



<details>

When `gteSupport` Feature flag is not enabled -

![image](https://user-images.githubusercontent.com/94359491/221147640-df159cec-e97f-4885-abd8-aec5caffb9ae.png)


When `gteSupport` Feature flag is enabled -

![image](https://user-images.githubusercontent.com/94359491/221147824-43940f74-ac37-4021-a393-a7c293865a9d.png)


https://user-images.githubusercontent.com/94359491/221149185-231a1422-0972-457c-bb0f-d4d53408d605.mp4




</details>
